src/NServiceBus.Heartbeat/HeartbeatSender.cs

```diff
@@ -23,41 +22,56 @@ class HeartbeatSender : FeatureStartupTask, IDisposable
             this.hostInfo = hostInfo;
         }
 
-        public void Dispose()
-        {
-            cancellationTokenSource?.Dispose();
-        }
+        public void Dispose() =>
+            stopSendingHeartbeatsTokenSource?.Dispose();
 
-        protected override Task OnStart(IMessageSession session)
+        protected override Task OnStart(IMessageSession session, CancellationToken cancellationToken = default)
         {
-            cancellationTokenSource = new CancellationTokenSource();
-
-            NotifyEndpointStartup(DateTime.UtcNow);
-            StartHeartbeats();
+            stopSendingHeartbeatsTokenSource = new CancellationTokenSource();
 
-            return Task.FromResult(0);
-        }
-
-        protected override Task OnStop(IMessageSession session)
-        {
-            heartbeatTimer?.Stop();
+            // don't block here since StartupTasks are executed synchronously.
+            _ = SendEndpointStartupMessage(DateTime.UtcNow, cancellationToken);
```

Comment:
I don't think we should pass the token here. `SendEndpointStartupMessage` runs recursively in a background task until it succeeds, after `OnStart` returns. If the token is cancelled after `OnStart` returns, `SendEndpointStartupMessage` must not be affected.

Reply:
Come to think of it... `OnStop()` should probably cancel that background task.

Reply:
Discussed with @DavidBoike and will make both changes.